File: wordle-rs/src/players.rs
//! This module provides methods and structs to manage information
//! about a Wordle player. Methods are also provided to save the
//! player to and load a player from a database file.
//!
//! Author: Benjamin Hall

use core::fmt;
use std::collections::HashSet;

use std::fmt::Write;
use std::fs::File;
use std::io::{self, BufReader, BufWriter, Read};

use crate::database::DatabaseEntry;

/// Contains information about a Wordle player.
///
/// A player has a:
/// - username
/// - list of words played
/// - guess distribution
/// - maximum win streak
/// - current win streak
pub struct PlayerInfo {
    username: String,
    words_played: HashSet<String>,
    num_guesses: [usize; 6],
    max_win_streak: usize,
    cur_win_streak: usize,
}

impl fmt::Display for PlayerInfo {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Username: {}", self.username)?;
        writeln!(
            f,
            "Words Played: {}",
            self.words_played
                .iter()
                .cloned()
                .collect::<Vec<_>>()
                .join(",")
        )?;
        writeln!(
            f,
            "Number of Guesses: {}",
            self.num_guesses.map(|i| i.to_string()).join(",")
        )?;
        writeln!(f, "Maximum Win Streak: {}", self.max_win_streak)?;
        writeln!(f, "Current Win Streak: {}", self.cur_win_streak)
    }
}

impl PlayerInfo {
    /// Initializes data for a new player.
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// ```
    pub fn new(username: String) -> Self {
        PlayerInfo::load(username, HashSet::new(), [0; 6], 0, 0)
    }

    /// Loads data for an existing player.
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let words_played = HashSet::new();
    /// words_played.insert(String::from("TRACE"));
    ///
    /// let num_guesses = [0; 6];
    /// num_guesses[4 - 1] += 1;
    ///
    /// let max_win_streak = 1;
    /// let cur_win_streak = 1;
    ///
    /// let player = PlayerInfo::new(
    ///     "user",
    ///     words_played,
    ///     num_guesses,
    ///     max_win_streak,
    ///     cur_win_streak
    /// );
    /// ```
    pub fn load(
        username: String,
        words_played: HashSet<String>,
        num_guesses: [usize; 6],
        max_win_streak: usize,
        cur_win_streak: usize,
    ) -> Self {
        PlayerInfo {
            username,
            words_played,
            num_guesses,
            max_win_streak,
            cur_win_streak,
        }
    }

    /// Gets the username of this player.
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// // prints "user"
    /// println!("{}", player.get_username());
    /// ```
    pub fn get_username(&self) -> &str {
        self.username.as_str()
    }

    /// Gets a random word this player has not yet played.
    ///
    /// # Examples:
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// let dictionary: HashSet<String> =
    ///     read_dictionary("dictionary.txt");
    /// let word = player.get_random_word(&dictionary);
    /// ```
    pub fn get_random_word(&self, dictionary: &HashSet<String>) -> String {
        let unplayed_words_cnt = dictionary.len() - self.words_played.len();
        let random_word_idx = fastrand::usize(0..unplayed_words_cnt);
        dictionary.iter().nth(random_word_idx).unwrap().to_owned()
    }

    /// Adds a word the player has successfully guessed to their database.
    ///
    /// This function increments the current win streak, adds
    /// the guess to the list of words played and to the guess
    /// distribution, and updates the max win streak if appropriate.
    ///
    /// # Examples:
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// // player got TRACE in 3 guesses
    /// player.add_won_word(String::from("TRACE"), 3);
    /// ```
    pub fn add_won_word(&mut self, word: String, num_guesses: usize) {
        self.words_played.insert(word);
        self.num_guesses[num_guesses - 1] += 1;
        self.cur_win_streak += 1;
        self.max_win_streak = std::cmp::max(self.max_win_streak, self.cur_win_streak);
    }

    /// Adds a word the player has failed to guess to their database.
    ///
    /// This function resets the current win streak to 0 and adds
    /// the guess to the list of words played. The number of guesses
    /// is not added to the player's guess distribution.
    ///
    /// # Examples:
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// // player could not guess BEBOP within 6 guesses
    /// player.add_lost_word(String::from("BEBOP"));
    /// ```
    pub fn add_lost_word(&mut self, word: String) {
        self.words_played.insert(word);
        self.cur_win_streak = 0;
    }

    /// Returns a string with formated player statistics.
    ///
    /// Player statistics consist of:
    /// - Number of words played
    /// - Win rate
    /// - Current and max win streak
    /// - Guess distribution
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// println!("{}", player.get_stats());
    /// ```
    pub fn get_stats(&self) -> String {
        let mut stats = String::new();
        writeln!(stats, "Number of Words Played: {}", self.words_played.len()).unwrap();

        let win_rate = if self.words_played.is_empty() {
            0
        } else {
            (100.0 * self.num_guesses.iter().sum::<usize>() as f64 / self.words_played.len() as f64)
                .round() as u32
        };
        writeln!(stats, "Win Rate: {win_rate}%").unwrap();

        writeln!(stats, "Current Win Streak: {}", self.cur_win_streak).unwrap();
        writeln!(stats, "Maximum Win Streak: {}", self.max_win_streak).unwrap();

        writeln!(stats, "Guess Distribution:").unwrap();
        let max_num_guesses = self.num_guesses.iter().max();

        /* print the guess distribution with bars, max 12 */
        let bar_factor = match max_num_guesses {
            Some(max_num_guesses) => 12.0 / *max_num_guesses as f64,
            None => 0.0,
        };

        for (i, num_guess) in self.num_guesses.iter().enumerate() {
            let num_bars = (bar_factor * *num_guess as f64).round() as usize;
            let bars = "=".repeat(num_bars);
            writeln!(stats, "{}: {bars} {num_guess}", i + 1).unwrap();
        }

        stats.trim().to_owned()
    }

    /// Writes this player's data to a file.
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::new("user");
    /// player.write_to_file("user.txt")?;
    /// ```
    pub fn write_to_file(&self, filename: &str) -> io::Result<()> {
        let file = File::create(filename)?;
        let mut writer = BufWriter::new(file);
        use std::io::Write;
        writer.write_all(self.to_string().as_bytes())
    }

    /// Reads a player's info from a file.
    ///
    /// All errors with reading the file, including any errors
    /// with parsing data, is propogated up to the caller.
    ///
    /// # Examples
    ///
    /// Basic usage:
    /// ```
    /// let player = PlayerInfo::from_file("user.txt")?;
    /// ```
    pub fn from_file(filename: &str) -> io::Result<Option<Self>> {
        let bad_data_err = || {
            io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Error: corrupt player database file: {}", filename),
            )
        };

        let file = File::open(filename);
        let file = match file {
            Ok(file) => file,
            Err(_) => return Ok(None),
        };

        let mut file_contents = String::new();
        let mut file_reader = BufReader::new(file);
        file_reader.read_to_string(&mut file_contents)?;
        /* read all the lines in the file */
        let lines_in_file: Vec<&str> = file_contents.lines().collect();

        if lines_in_file.len() != 5 {
            /* corrupt database file */
            return Err(bad_data_err());
        }

        /* parse the lines in the file */
        let username = DatabaseEntry::from_line(lines_in_file[0], str::to_owned);
        let words_played = DatabaseEntry::from_set(lines_in_file[1], str::to_owned);
        let num_guesses_list = DatabaseEntry::from_list(lines_in_file[2], str::parse::<usize>);
        let max_win_streak = DatabaseEntry::from_line(lines_in_file[3], str::parse::<usize>);
        let cur_win_streak = DatabaseEntry::from_line(lines_in_file[3], str::parse::<usize>);

        /* make sure no data was missing */
        if username.is_none()
            || words_played.is_none()
            || num_guesses_list.is_none()
            || max_win_streak.is_none()
            || cur_win_streak.is_none()
        {
            return Err(bad_data_err());
        }
        let username = username.unwrap();
        let words_played = words_played.unwrap();
        let num_guesses_list = num_guesses_list.unwrap();
        let max_win_streak = max_win_streak.unwrap();
        let cur_win_streak = cur_win_streak.unwrap();

        /* propogate any str::parse() errors to the caller */
        let num_guesses_list = DatabaseEntry {
            name: num_guesses_list.name,
            value: num_guesses_list
                .value
                .into_iter()
                .map(|r| r.map_err(|_| bad_data_err()))
                .collect::<Result<Vec<_>, _>>()?,
        };
        let max_win_streak = DatabaseEntry {
            name: max_win_streak.name,
            value: max_win_streak.value.map_err(|_| bad_data_err())?,
        };
        let cur_win_streak = DatabaseEntry {
            name: cur_win_streak.name,
            value: cur_win_streak.value.map_err(|_| bad_data_err())?,
        };

        /* parse the number of guesses into an array */
        let mut num_guesses = [0; 6];
        num_guesses_list
            .value
            .iter()
            .take(num_guesses.len())
            .enumerate()
            .for_each(|(i, num_guess)| num_guesses[i] = *num_guess);

        let player = PlayerInfo::load(
            username.value,
            words_played.value,
            num_guesses,
            max_win_streak.value,
            cur_win_streak.value,
        );
        Ok(Some(player))
    }
}

File: WordleCpp/src/cpp/players/players.cpp
/*
 * Wordle program
 * Author: Benjamin Hall
 */

#include "players/players.h"
#include "players/database.h"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <numeric>
#include <sstream>
#include <vector>

namespace players {

PlayerInfo::PlayerInfo(std::string username)
{
    m_username = std::move(username);
}

PlayerInfo::PlayerInfo(std::string username,
                    std::unordered_set<std::string> wordsPlayed,
                    std::array<uint32_t, MAX_NUM_GUESSES> numGuesses,
                    uint32_t maxWinStreak,
                    uint32_t curWinStreak)
{
    m_username = std::move(username);
    m_wordsPlayed = std::move(wordsPlayed);
    m_numGuesses = std::move(numGuesses);
    m_maxWinStreak = maxWinStreak;
    m_curWinStreak = curWinStreak;
}

std::string const &PlayerInfo::GetUsername() const
{
    return m_username;
}

std::string PlayerInfo::GetRandomWord(std::unordered_set<std::string> const &dictionary) const
{
    size_t unplayedWords = dictionary.size() - m_wordsPlayed.size();
    size_t randomWordIdx = rand() % unplayedWords;

    auto iter = dictionary.begin();
    for (int i = 0; i < randomWordIdx; ++i) {
        ++iter;
    }
    return *iter;
}

void PlayerInfo::AddWonWord(std::string const &word, uint32_t numGuesses)
{
    m_wordsPlayed.insert(word);
    ++m_numGuesses[numGuesses - 1];
    ++m_curWinStreak;
    m_maxWinStreak = std::max(m_maxWinStreak, m_curWinStreak);
}

void PlayerInfo::AddLostWord(std::string const &word)
{
    m_wordsPlayed.insert(word);
    m_curWinStreak = 0;
}

std::string PlayerInfo::ToString() const
{
    std::stringstream playerStr;
    playerStr << "Username: " << m_username << std::endl;

    playerStr << "Words Played: ";
    {
        bool firstElem = true;
        for (auto const &wordPlayed : m_wordsPlayed) {
            if (firstElem) {
                firstElem = false;
            } else {
                playerStr << ",";
            }
            playerStr << wordPlayed;
        }
        playerStr << std::endl;
    }

    playerStr << "Number of Guesses: "
        << std::accumulate(m_numGuesses.begin(), m_numGuesses.end(), 0, std::plus<uint32_t>())
        << std::endl;

    playerStr << "Maximum Win Streak: " << m_maxWinStreak << std::endl;
    playerStr << "Current Win Streak: " << m_curWinStreak << std::endl;

    return playerStr.str();
}

std::string PlayerInfo::GetStats() const
{
    std::stringstream playerStats;

    playerStats << "Number of Words Played: " << m_wordsPlayed.size() << std::endl;

    playerStats << "Win Rate: " << [&] {
            int wonGames = std::accumulate(m_numGuesses.begin(), m_numGuesses.end(), 0, std::plus<uint32_t>());
            return (uint32_t)std::round(100.0 * wonGames / m_wordsPlayed.size());
        }() << std::endl;
    
    playerStats << "Current Win Streak: " << m_curWinStreak << std::endl;
    playerStats << "Maximum Win Streak: " << m_maxWinStreak << std::endl;
    playerStats << "Guess Distribution:";

    double barFactor = [&] {
        auto maxNumGuesses = std::max_element(m_numGuesses.begin(), m_numGuesses.end());
        if (maxNumGuesses == m_numGuesses.end() || *maxNumGuesses == 0) {
            return 0.0;
        } else {
            return 12.0 / *maxNumGuesses;
        }
    }();

    for (int i = 0; i < m_numGuesses.size(); ++i) {
        int numBars = (int)std::round(barFactor * m_numGuesses[i]);
        playerStats << std::endl << (i + 1) << ": " << std::string(numBars, '=') << m_numGuesses[i];
    }

    return playerStats.str();
}

int PlayerInfo::WriteToFile(std::string const &filename) const
{
    std::ofstream file{filename};
    if (!file.is_open()) {
        return -1;
    }

    file << ToString();
    file.close();

    return 0;
}

/* static */ std::optional<PlayerInfo> PlayerInfo::CreateFromFile(std::string const &filename)
{
    std::ifstream file{filename};
    if (!file.is_open()) {
        return std::nullopt;
    }

    static std::string const errStr = "Error: corrupt player database file: " + filename;

    std::string line;

    if (!std::getline(file, line)) {
        throw errStr;
    }
    auto username = DatabaseEntry<std::string>::FromLine(line, [](auto const &s) { return s; });
    if (!username.has_value()) {
        throw errStr;
    }

    if (!std::getline(file, line)) {
        throw errStr;
    }
    auto wordsPlayed = DatabaseEntry<std::string>::FromHashset(line, [](auto const &s) { return s; });
    if (!wordsPlayed.has_value()) {
        throw errStr;
    }

    if (!std::getline(file, line)) {
        throw errStr;
    }
    auto numGuessesList = DatabaseEntry<uint32_t>::FromVector(line, [](auto const &s) { return std::stol(s); });
    if (!numGuessesList.has_value()) {
        throw errStr;
    }

    if (!std::getline(file, line)) {
        throw errStr;
    }
    auto maxWinStreak = DatabaseEntry<uint32_t>::FromLine(line, [](auto const &s) { return std::stol(s); });
    if (!maxWinStreak.has_value()) {
        throw errStr;
    }

    if (!std::getline(file, line)) {
        throw errStr;
    }
    auto curWinStreak = DatabaseEntry<uint32_t>::FromLine(line, [](auto const &s) { return std::stol(s); });
    if (!curWinStreak.has_value()) {
        throw errStr;
    }

    file.close();

    std::array<uint32_t, MAX_NUM_GUESSES> numGuesses{};
    for (int i = 0; i < numGuesses.size() && i < numGuessesList->value.size(); ++i) {
        numGuesses[i] = numGuessesList->value[i];
    }

    return std::optional{
            PlayerInfo{std::move(username->value), std::move(wordsPlayed->value),
                    std::move(numGuesses), maxWinStreak->value, curWinStreak->value}
        };
}

}
